src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -1,100 +1,312 @@
+
+                return null;
+            }
+        }
+
+        private class IncrementalAnalyzer : IncrementalAnalyzerBase
+        {
+            private readonly ConcurrentDictionary<ProjectId, ProjectInfo> _projectToInfo;
+            private readonly ConcurrentDictionary<string, MetadataInfo> _metadataPathToInfo;
+
+            private readonly Dictionary<ProjectId, Project> _idTolastSeenProject = new Dictionary<ProjectId, Project>();
+
+            public IncrementalAnalyzer(
+                ConcurrentDictionary<ProjectId, ProjectInfo> projectToInfo,
+                ConcurrentDictionary<string, MetadataInfo> metadataPathToInfo)
+            {
+                _projectToInfo = projectToInfo;
+                _metadataPathToInfo = metadataPathToInfo;
             }
 
-            public override Task NewSolutionSnapshotAsync(Solution solution, CancellationToken cancellationToken)
+            public override Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt, CancellationToken cancellationToken)
             {
-                // check whether we are good to report total symbol numbers
-                if (_symbolCountByProjectMap == null || _symbolCountByProjectMap.Count < solution.ProjectIds.Count || string.IsNullOrEmpty(solution.FilePath))
+                if (!document.SupportsSyntaxTree)
                 {
+                    // Not a language we can produce indices for (i.e. TypeScript).  Bail immediately.
                     return SpecializedTasks.EmptyTask;
                 }
 
-                if (_solutionId != null && _solutionId != solution.Id)
+                if (bodyOpt != null)
                 {
-                    ReportCount();
+                    // This was a method level edit.  This can't change the symbol tree info
+                    // for this project.  Bail immediately.
                     return SpecializedTasks.EmptyTask;
                 }
 
-                _solutionId = solution.Id;
-                foreach (var projectId in solution.ProjectIds)
+                return UpdateSymbolTreeInfoAsync(document.Project, cancellationToken);
+            }
+            
+            public override Task AnalyzeProjectAsync(Project project, bool semanticsChanged, CancellationToken cancellationToken)
+            {
+                return UpdateSymbolTreeInfoAsync(project, cancellationToken);
+            }
+
+            private async Task UpdateSymbolTreeInfoAsync(Project project, CancellationToken cancellationToken)
+            {
+                if (!project.SupportsCompilation)
                 {
-                    if (!_symbolCountByProjectMap.ContainsKey(projectId))
-                    {
-                        return SpecializedTasks.EmptyTask;
-                    }
+                    return;
+                }
+
+                Project lastSeenProject;
+                if (_idTolastSeenProject.TryGetValue(project.Id, out lastSeenProject) && lastSeenProject == project)
```

Comment:
this is actually not true that project will always be from current solution, eventually it will be, but while user typing it is possible, temporarily we hold onto project from multiple solutions. I think this logic can be changed to hold onto project version instead of project itself. and rather check version to see whether it already processed the project.